### PR TITLE
su: Add --login option + LibCore: More error propagation

### PR DESCRIPTION
**Pull request by LucasChollet**
This PR includes the --login option for `su` and more error propagation that I wrote along.

### CODE REVIEW

File: Userland/Utilities/su.cpp [hunk 1]
```diff
@@ -21,37 +21,50 @@ ErrorOr<int> serenity_main(Main::Arguments arguments)
...
+    if (first_positional == '-') {
+        simulate_login = true;
+        user = second_positional;
+    }
+
     if (geteuid() != 0)
         return Error::from_string_literal("Not running as root :(");
 
-    auto account = TRY(user ? Core::Account::from_name(user) : Core::Account::from_uid(0));
+    auto account = TRY(user.is_null() ? Core::Account::from_uid(0) : Core::Account::from_name(user));
```
Comment (AtkinsSJ) [inline Userland/Utilities/su.cpp:44]: `String::is_null()` and `StringView::is_null()` are things we're trying to replace with `Optional<>` or just "empty" strings. `ArgsParser` seems to not yet support giving it an `Optional<StringView>`, so I'd suggest initializing `first_positional` and `second_positional` to `""sv` and then using `is_empty()`. (Unless you feel like adding the support to ArgsParser. :yakbait:)

File: Userland/Utilities/su.cpp [hunk 2]
```diff
@@ -21,37 +21,50 @@ ErrorOr<int> serenity_main(Main::Arguments arguments)
...
+
     TRY(Core::System::pledge("stdio exec"));
 
     TRY(Core::System::setenv("HOME"sv, account.home_directory(), true));
 
-    execl(account.shell().characters(), account.shell().characters(), nullptr);
-    perror("execl");
-    return 1;
+    TRY(Core::System::exec(account.shell().characters(), Array<StringView, 1> { account.shell().view() }, Core::System::SearchInPath::No));
+    return { 1 };
```
Comment (AtkinsSJ) [inline Userland/Utilities/su.cpp:66]: `return 1;` should be fine here.

**Review by AtkinsSJ — CHANGES_REQUESTED** [review]

File: AK/StringView.h [hunk 3]
```diff
@@ -211,6 +211,11 @@ class StringView {
         return *cp == '\0';
     }
 
+    constexpr bool operator==(char const c) const
```
[the remark this answers is not in the document]
Comment (LucasChollet) [reply AK/StringView.h:214]: As you pointed out, as it is passed by value, it won't modify the `char` of the caller. This `const` means that the stack variable `c` won't be modified in this function, i.e. `c` won't be assigned to anything else during its lifetime.

File: Userland/Utilities/su.cpp (hunk 1 above)
Comment (LucasChollet) [reply Userland/Utilities/su.cpp:44]: I will change it to use `is_empty()`.
I think that this PR already has enough side commits, but I will have a look at `ArgsParser` :yakplus:

File: Userland/Libraries/LibCore/System.cpp [hunk 4]
```diff
@@ -1205,6 +1198,13 @@ ErrorOr<Vector<gid_t>> getgroups()
     return groups;
 }
 
+ErrorOr<void> setgroups(Span<gid_t const> gids)
```
Comment (bgianfo) [inline Userland/Libraries/LibCore/System.cpp:1201]: Please don't move code, if you are going to move code put it in its own commit, not mixes with other functional changes.

Comment (bgianfo) [reply Userland/Libraries/LibCore/System.cpp:1201]: That includes moving it out of ifdefs 😁

**Comment by linusg** [comment]
This has a conflict

File: Userland/Libraries/LibCore/Account.cpp [hunk 5]
```diff
@@ -152,7 +152,7 @@ bool Account::login() const
         return false;
 
     auto const temporary_directory = String::formatted("/tmp/{}", m_uid);
-    if (auto result = Core::Directory::create(temporary_directory, Core::Directory::CreateDirectories::No); result.is_error())
+    if (auto result = Core::Directory::create(temporary_directory, Core::Directory::CreateDirectories::Yes); result.is_error())
         dbgln("{}", result.release_error());
```
Comment (awesomekling) [inline Userland/Libraries/LibCore/Account.cpp]: Hold up, what if another malicious user has already created this directory?

Comment (LucasChollet) [reply Userland/Libraries/LibCore/Account.cpp]: You're right, we need something more secure.

Maybe we can create a directory at first login and store the value in a static map in `Account`. It's default value would be the current value (just the UID) and if it's taken we iterate on UID-X until we find something. Then further call of `parse_uid_in_path` will restore the value from the map.

Comment (awesomekling) [reply Userland/Libraries/LibCore/Account.cpp]: How about we create the directory somewhere in the user's home directory instead? Is there some reason we have to use the shared playground in `/tmp`?

Comment (LucasChollet) [reply Userland/Libraries/LibCore/Account.cpp]: The only issue I can think of (it also exists when using /tmp) is stale sockets living after the crash of the owner.

Using a path outside /tmp, prevent a magic reboot to solve the issue.

I'm not very familiar with Unix sockets and even less with serenity's one.  I could definitely have missed something.

Comment (LucasChollet) [reply Userland/Libraries/LibCore/Account.cpp]: From what I read [here](https://stackoverflow.com/questions/42555403/unix-domain-socket-file-not-removed-on-application-crash), I suppose that it would work.

Abstract socket might also be worth implementing.

Anyway, I will move it to the user's personal directory and check that leftover portals are indeed cleaned before creation.

Comment (LucasChollet) [reply Userland/Libraries/LibCore/Account.cpp]: I finally moved this commit to #14673.

File: Userland/Services/LoginServer/main.cpp [hunk 6]
```diff
@@ -99,13 +99,13 @@ ErrorOr<int> serenity_main(Main::Arguments arguments)
...
 
-    char const* auto_login = nullptr;
+    StringView auto_login;
 
     Core::ArgsParser args_parser;
     args_parser.add_option(auto_login, "automatically log in with no prompt", "auto-login", 'a', "username");
     args_parser.parse(arguments);
 
-    if (!auto_login) {
+    if (auto_login.is_null()) {
```
Comment (AtkinsSJ) [inline Userland/Services/LoginServer/main.cpp:108]: Use `is_empty()` instead.

**Review by AtkinsSJ — CHANGES_REQUESTED** [review]
Wow, this has been hanging around for a while! :sweat_smile:

Overall it's good, but can you please modify `/Base/usr/share/man/man1/su.md` to mention your new options?

File: Userland/Services/LoginServer/main.cpp (hunk 6 above)
Comment (LucasChollet) [reply Userland/Services/LoginServer/main.cpp:108]: Sorry, I should have pay attention to that, you already mentioned it to me!

**Comment by LucasChollet** [comment]
It now includes an Options section in `su.md`!
